http-clients/netty-nio-client/src/test/java/software/amazon/awssdk/http/nio/netty/internal/GoAwayTest.java

```diff
@@ -0,0 +1,331 @@
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.CountDownLatch;
+import java.util.function.Supplier;
+import org.junit.After;
+import org.junit.Test;
+import org.reactivestreams.Publisher;
+import software.amazon.awssdk.http.Protocol;
+import software.amazon.awssdk.http.SdkHttpFullRequest;
+import software.amazon.awssdk.http.SdkHttpMethod;
+import software.amazon.awssdk.http.SdkHttpResponse;
+import software.amazon.awssdk.http.async.AsyncExecuteRequest;
+import software.amazon.awssdk.http.async.SdkAsyncHttpClient;
+import software.amazon.awssdk.http.async.SdkAsyncHttpResponseHandler;
+import software.amazon.awssdk.http.nio.netty.EmptyPublisher;
+import software.amazon.awssdk.http.nio.netty.NettyNioAsyncHttpClient;
+import software.amazon.awssdk.http.nio.netty.SdkEventLoopGroup;
+
+/**
+ * Tests to ensure that the client behaves as expected when it receives GOAWAY messages.
+ */
+public class GoAwayTest {
+
+    private SdkAsyncHttpClient netty;
+    private SimpleEndpointDriver endpointDriver;
+
+    @After
+    public void teardown() throws InterruptedException {
+        if (endpointDriver != null) {
+            endpointDriver.shutdown();
+        }
+        endpointDriver = null;
+
+        if (netty != null) {
+            netty.close();
+        }
+        netty = null;
+    }
+
+    @Test
+    public void execute_goAwayReceived_existingChannelsNotReused() throws InterruptedException {
+        // Frame listener supplier for each connection
+        Supplier<Http2FrameListener> frameListenerSupplier = () -> new TestFrameListener() {
+            @Override
+            public void onHeadersRead(ChannelHandlerContext ctx, int streamId, Http2Headers headers, int streamDependency, short weight, boolean exclusive, int padding, boolean endStream) {
+                frameWriter().writeHeaders(ctx, streamId, new DefaultHttp2Headers().add("content-length", "0").status("204"), 0, true, ctx.newPromise());
+                ctx.flush();
+            }
+        };
+
+        endpointDriver = new SimpleEndpointDriver(frameListenerSupplier);
+        endpointDriver.init();
+
+        netty = NettyNioAsyncHttpClient.builder()
+                .eventLoopGroup(SdkEventLoopGroup.builder().numberOfThreads(1).build())
+                .protocol(Protocol.HTTP2)
+                .build();
+
+        sendGetRequest().join();
+
+        // Note: It's possible the initial request can cause the client to allocate more than 1 channel
+        int initialChannelNum = endpointDriver.channels.size();
+
+        // Send GOAWAY to all the currently open channels
+        endpointDriver.channels.forEach(ch -> endpointDriver.goAway(ch, 1));
+
+        // Need to give a chance for the streams to get closed
+        Thread.sleep(1000);
```

Comment:
Have we run this enough to be convinced it isn't flaky?

Reply:
1s is a pretty long time to close 2 connections so it should be fine